### PR TITLE
Send2Ebook (receiver) plugin

### PR DESCRIPTION
**Pull request by mwoz123**
## Decription
Send2Ebook lets you send articles found on PC/Android phone to your Ebook reader.

## How it works
Open your faviourite Android web browser, then `Share via`, choose `Send2Ebook`.
On your Ebook reader download articles

## links
https://github.com/koreader/android-send2ebook
core lib (console UI):
https://github.com/mwoz123/send2ebook

### CODE REVIEW

File: frontend/apps/cloudstorage/ftp.lua [hunk 1]
```diff
@@ -8,30 +8,20 @@ local UIManager = require("ui/uimanager")
...
-        at_sign = "@"
-    end
-    if pass ~= "" then
-        colon_sign = ":"
-    end
-    local replace = "://" .. user .. colon_sign .. pass .. at_sign
-    local url = string.gsub(address, "://", replace)
-    return url
-end
+
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/ftp.lua]: Stray newline

File: plugins/send2ebook.koplugin/ftpconnectionconfigfile.lua [hunk 2]
```diff
@@ -0,0 +1,10 @@
+return {
+
+--hint: host needs to start with 'ftp://' prefix otherwise won't work
+
+	host='ftp://server.com',
```
Comment (Frenzie) [inline plugins/send2ebook.koplugin/ftpconnectionconfigfile.lua]: tab

File: plugins/send2ebook.koplugin/main.lua [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+local initialized = false
+local wifi_enabled_before_action = true
+local ftp_connection_config_file = "ftpconnectionconfig.lua"
+local send_to_koreader_config_file = "send_to_koreader_settings.lua"
+local config_key_custom_dl_dir = "custom_dl_dir";
+local default_download_dir_name = "Send2Ebook"
+local download_dir_path
+
+local function stringEnds(String,End)
```
Comment (Frenzie) [inline plugins/send2ebook.koplugin/main.lua]: variables should be lowercase

File: plugins/send2ebook.koplugin/main.lua [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+   return End=='' or string.sub(String,-string.len(End))==End
```
Comment (Frenzie) [inline plugins/send2ebook.koplugin/main.lua]: Please use `""`

File: plugins/send2ebook.koplugin/main.lua [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+end
+
+function Send2Ebook:downloadFileAndRemove(connectionUrl, remotePath, localDownloadPath)
```
Comment (Frenzie) [inline plugins/send2ebook.koplugin/main.lua]: connection_url, remote_path and local_download_path

File: plugins/send2ebook.koplugin/main.lua [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+        if not lfs.attributes(download_dir_path, "mode") then
+            logger.dbg("Send2Ebook: Creating initial directory")
+            lfs.mkdir(download_dir_path)
+        end
+    end
+end
+
+function Send2Ebook:process()
+    local info = InfoMessage:new{ text = _("Connecting ...") }
```
Comment (Frenzie) [inline plugins/send2ebook.koplugin/main.lua]: `Connecting…`

**Comment by mwoz123** [comment]
Thanks. Will fix them later. 

One question :
Is there a method to check if ebook has been opened? Is it :`DocSettings:getSidecarFile(file_abs_path)`? 

As I used this plugin for a while and I think it makes more sense to "delete only those which were opened" instead 'delete all'. 
Many times I had plenty of unreaded z (and plenty of readed) and would like to clean the mess before downloading new

**Comment by Frenzie** [comment]
has if it's just to check if it exists

http://koreader.rocks/doc/modules/docsettings.html#DocSettings:hasSidecarFile

**Comment by mwoz123** [comment]
>Why not simply os.remove?

I refactored it a bit. and `os.remove` doesn't work for me.


Is there a simply way to create editable configuration UI for entering:
- host, username, password, folder

or it has to be tons of code like in cloudstorage?

Maybe if it not too complicated I'll be able to replace the manual edit ?

**Comment by Frenzie** [comment]
What tons of code are you referring to? It looks pretty minimal to me.

https://github.com/koreader/koreader/blob/aadb81eb4831c968daf14e1c20845c9810bb5f6e/frontend/apps/cloudstorage/ftp.lua#L57-L149

**Comment by mwoz123** [comment]
> What tons of code are you referring to? It looks pretty minimal to me.

code looked complicated. But I manage to code it.

**Comment by Frenzie** [comment]
> code looked complicated.

UI is always complicated, although I think [QML](http://doc.qt.io/qt-5/qmlapplications.html) is more elegant than most attempts at making it easier to work with. But I don't think it's really any better or worse than Gtk, Tcl/Tk, Java's Swing and its successor JavaFX, etc. :-)

File: plugins/send2ebook.koplugin/main.lua [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    UIManager:show(info)
+    Send2Ebook:afterWifiAction()
+end
+
+function Send2Ebook:removeReadActicles()
+    logger.dbg("Send2Ebook: Removing read articles from :", download_dir_path)
+    for entry in lfs.dir(download_dir_path) do
+        if entry ~= "." and entry ~= ".." then
+            local entry_path = download_dir_path .. entry
+            if DocSettings:hasSidecarFile(entry_path) then
```
Comment (mwoz123) [inline plugins/send2ebook.koplugin/main.lua]: Does this pr https://github.com/koreader/koreader/pull/3675/files
Make any influence to above check? Cause after merge master it doesn't seem to work correctly, which it did on emulator (and old master)

Comment (mwoz123) [reply plugins/send2ebook.koplugin/main.lua]: Maybe I should use something else here?

Comment (mwoz123) [reply plugins/send2ebook.koplugin/main.lua]: It seems to be working fine on emulator, after todays master merge

**Comment by mwoz123** [comment]
any comments? merge?

**Comment by Frenzie** [comment]
Oops, I had a pending comment on the question that I forgot to submit. But that seems to be no longer an issue. 

> No, it's the same way files are displayed read or unread in the filemanager. What are the steps?

I'll take a look. It should be ready or just one or two dotted is from it.

File: plugins/send2ebook.koplugin/main.lua [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+function Send2Ebook:process()
+    local info = InfoMessage:new{ text = _("Connecting …") }
+    UIManager:show(info)
+    logger.dbg("Send2Ebook: force repaint due to upcoming blocking calls")
+    UIManager:forceRePaint()
+    UIManager:close(info)
+
+    local count = 1
+    local ftp_config = send2ebook_settings:readSetting("ftp_config") or {address='ftp://Please_setup_ftp_settings_first;)', username="", password="", folder=""}
```
Comment (Frenzie) [inline plugins/send2ebook.koplugin/main.lua]: Is `ftp://Please_setup_ftp_settings_first;)` a typo of some sort? (Particularly that `;)`)

Also double quotes would be nice. ;-)

File: plugins/send2ebook.koplugin/main.lua [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+    local connection_url = FtpApi:generateUrl(ftp_config.address, ftp_config.username, ftp_config.password)
+
+    local fileTable = FtpApi:listFolder(connection_url .. ftp_config.folder, ftp_config.folder) --args looks strange but otherwise resonse with invalid paths
```
Comment (Frenzie) [inline plugins/send2ebook.koplugin/main.lua]: It's not a function.

File: plugins/send2ebook.koplugin/main.lua [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+    if not fileTable then
+      info = InfoMessage:new{ text = T(_("Could not get file list for server: %1, user: %2, folder: %3"), ftp_config.address, ftp_config.username, ftp_config.folder) }
+    else
+      local total_entries = table.getn(fileTable)
+      logger.dbg("total_entries ", total_entries)
+      if total_entries > 1 then total_entries = total_entries -2 end --remove result "../" (upper folder) and "./" (current folder)
+      for idx, ftpFile in ipairs(fileTable) do
```
Comment (Frenzie) [inline plugins/send2ebook.koplugin/main.lua]: ftp_file is also not a function

**Comment by Frenzie** [comment]
Thanks!

**Pull request merged**